src/circuit.cpp

```diff
@@ -1,13 +1,331 @@
+block_type hash_layer(std::array<block_type, LayerSize> input, size_t layer){
+    constexpr size_t NextLayerSize = (LayerSize % 2 == 0) ? LayerSize / 2 : (LayerSize / 2 + 1);
+    std::array<block_type, NextLayerSize> next_layer;
+
+    for (std::size_t leaf_index = 0; leaf_index < LayerSize / 2; leaf_index++) {
+        next_layer[leaf_index] = hash<hash_type>(input[2 * leaf_index], input[2 * leaf_index + 1]);
+    }
+    if (LayerSize % 2 != 0) {
+        next_layer[NextLayerSize - 1] = hash<hash_type>(LayerSize - 1, precomputed_zero_hashes[layer]);
+    }
+    if (LayerSize == 2)
+        return next_layer[0];
+    else
+        return hash_layer<NextLayerSize>(next_layer, layer + 1);
+}
+
+template <std::size_t LayerSize>
+block_type hash_tree(std::array<block_type, LayerSize> input){
+    return hash_layer(input, 0);
+}
+
+template <std::size_t BalancesCount>
+std::array<block_type, BalancesCount / BALANCES_PER_LEAF> pack_balances_into_field_elements(
+    std::array<uint64_t, BalancesCount> validator_balances
+) {
+    std::array<block_type, BALANCES_LEAFS_COUNT> leafs;
+    for (std::size_t i = 0; i < BALANCES_LEAFS_COUNT; i++) {
+        // MSB first
+        decomposed_int64_type first_balance_in_block_bits =
+            __builtin_assigner_bit_decomposition64(validator_balances[4*i], BYTE_ORDER_LSB);
+        decomposed_int64_type second_balance_in_block_bits =
+            __builtin_assigner_bit_decomposition64(validator_balances[4*i+1], BYTE_ORDER_LSB);
+        decomposed_int64_type third_balance_in_block_bits =
+            __builtin_assigner_bit_decomposition64(validator_balances[4*i+2], BYTE_ORDER_LSB);
+        decomposed_int64_type fourth_balance_in_block_bits =
+            __builtin_assigner_bit_decomposition64(validator_balances[4*i+3], BYTE_ORDER_LSB);
+
+        typename field_type::value_type first_block = __builtin_assigner_bit_composition128(
+            first_balance_in_block_bits, second_balance_in_block_bits, BYTE_ORDER_LSB);
+        typename field_type::value_type second_block = __builtin_assigner_bit_composition128(
+            third_balance_in_block_bits, fourth_balance_in_block_bits, BYTE_ORDER_LSB);
+
+        leafs[i] = {first_block, second_block};
+    }
+    return leafs;
+}
+
+block_type lift_uint64(uint64_t value) {
+    return {
+        __builtin_assigner_bit_composition128(
+            __builtin_assigner_bit_decomposition64(value, BYTE_ORDER_LSB), 0, BYTE_ORDER_LSB
+        ),
+        0
+    };
+}
+
+block_type mix_in_size(const block_type root, size_t size) {
+    return hash<hash_type>(root, lift_uint64(size));
+}
+
+template <size_t TargetTreeHeight, size_t LeafsCount, size_t LeafsTreeHeight>
+block_type merkelize(std::array<block_type, LeafsCount> merkle_leaves) {
+    block_type leaves_tree_root = hash_tree<LeafsCount>(merkle_leaves);
+
+    block_type current_hash = leaves_tree_root;
+    for (size_t height = LeafsTreeHeight; height < TargetTreeHeight; ++height) {
```

Comment:
This also creates a bifurcation point gif execution trace - although it changes "slower"; not with every actual_validators count, but with every log2(actual_validators count)